prime-router/docs/features/0004-managing-translation-schemas.md

```diff
@@ -0,0 +1,217 @@
+
+### Azure blob storage (Recommended)
+This solution would maintain the majority of the existing functionality and just provide updates to use the azure blob storage as the underlying "file" storage rather than disk.  This solution would get versioning automatically by enabling [azure blob versioning](https://learn.microsoft.com/en-us/azure/storage/blobs/versioning-overview); this functionality enables automatically creating new versions when uploading a new file and then APIs for rolling back file updates in the event that there was a problem.
+
+Additionally, the versioning has the added flexibility of referencing a specific version of a blob so schemas could reference version 3 while the next version is iterated on.
+`https://<storage-account>.blob.core.windows.net/<container>/<blob-name>?versionid=<version-id>` . If the decision is made to continue to store some of the common schemas in source code this is an approach that could be adopted here as well by adding a version to the directory paths `file:///{BASE_DIR}/metadata/hl7_mapping/common/patient/v1/patient.yml``
+
+#### FHIR -> FHIR, FHIR -> HL7
+Steps for migrating:
+
+1. Update the code to use the URI to determine if the schema should be read from disk or azure by looking at the URI scheme
+    - Fallback to the existing behavior if it is not a valid URI (i.e. we're resolving an old schema with relative paths)
+2. Update all the schema references (i.e. `extends`, `schemaRef`, `schemaName`, etc.) in the existing schemas and settings to reference absolute paths as URLs
+    - `schema: ../common/patient` -> `schema: file:///{BASE_DIR}/metadata/hl7_mapping/common/patient.yml`
+3. Update schemas to reference azure blobs (i.e `azure:///{AZURE_STORAGE_LOCATION}/metadata/hl7_mapping/common/patient.yml`) and upload schemas to azure blob service
+4. Update all UP sender and receivers to reference the azure schema
+
+During the migration process, it would make sense to all take another look at the directory structure and names to make sure they make sense moving forward.
+
+#### HL7 -> FHIR
+Long term, it would be great to add support to this library (it's open sourced) that would support reading files from different file-like storage solutions rather than just disk, but this would likely not be feasible to get done in a quick enough timeframe.  Instead, before instantiating the converter, the application will sync the azure storage to a spot on the disk the library will read from.
+
+#### Cache
+**This could potentially be deferred until it's been identified that the network requests to azure blob storage are the bottleneck**
+
+Since the pipeline will now require reading several files out of the blob store at multiple spots, it will likely make sense to cache the resolved schemas since they will change relatively infrequently.  This could initially be done with a simple in-memory cache using the schema URI as the cache key.
```

Comment:
I left this relatively open ended since I think we would want to evaluate if there is a performance bottle neck here that needs addressing before implementing a cache.

There is the simplest case where there is just an in-memory map that exists as long as azure maintains the function instance; they're destroyed after [five minutes](https://learn.microsoft.com/en-us/azure/azure-functions/functions-scale#timeout) (I think we're on the consumption plan), but will be re-used if they it keeps getting invoked.  The invalidation section I have below would not apply for this implementation

If this was not sufficient or there was a desire for more fine-grained control, something like [azure cache service](https://azure.microsoft.com/en-us/products/cache) could be used

Reply:
Are files cached in memory right now? I've had to restart the app every time I make a change to the mappings.

Reply:
They're not, but they do get copied over into the `build` directory.  You could likely edit them there to get live updates.